### PR TITLE
Add zopfli extension

### PR DESCRIPTION
**Pull request by toukirkhan**
I have created a new PR due to merge conflicts, please review. Suggestions appreciated.

Reference:
https://github.com/appwrite/appwrite/issues/4447

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -109,6 +121,7 @@ COPY --from=brotli /usr/local/lib/php/extensions/no-debug-non-zts-20200930/brotl
 COPY --from=lz4 /usr/local/lib/php/extensions/no-debug-non-zts-20200930/lz4.so /usr/local/lib/php/extensions/no-debug-non-zts-20200930/
 COPY --from=snappy /usr/local/lib/php/extensions/no-debug-non-zts-20200930/snappy.so /usr/local/lib/php/extensions/no-debug-non-zts-20200930/
 COPY --from=xz /usr/local/lib/php/extensions/no-debug-non-zts-20200930/xz.so /usr/local/lib/php/extensions/no-debug-non-zts-20200930/
+COPY --from=zopfli /usr/local/lib/php/extensions/no-debug-non-zts-20200930/zopfli.so/usr/local/lib/php/extensions/no-debug-non-zts-20200930/
```
Comment (stnguyen90) [inline Dockerfile:124]: typo:

```suggestion
COPY --from=zopfli /usr/local/lib/php/extensions/no-debug-non-zts-20200930/zopfli.so /usr/local/lib/php/extensions/no-debug-non-zts-20200930/
```

**Review by stnguyen90 — CHANGES_REQUESTED** [review]

**Comment by stnguyen90** [comment]
Please also make sure the linter passes 🙏🏼

**Review by eldadfux — APPROVED** [review]

**Comment by eldadfux** [comment]
@toukirkhan we would love to merge this and include it in our next release, but we need tests to pass first.

**Comment by toukirkhan** [comment]
Sure, I will try to make the changes @eldadfux

**Comment by toukirkhan** [comment]
@eldadfux I feel that there's some file missing in the php extension of the zopfli compression am using due to which the build is failing.

File: Dockerfile [hunk 2]
```diff
@@ -86,6 +88,14 @@ RUN git clone https://github.com/codemasher/php-ext-xz.git --branch ${PHP_EXT_XZ
   && phpize \
   && ./configure \
   && make && make install
+  
+## Zopfli Extension
+FROM compile AS zopfli
+RUN git clone https://github.com/artfulhacker/php_zopfli.git \
```
Comment (stnguyen90) [inline Dockerfile:94]: This extension is really old. Let's use [this](https://github.com/kjdev/php-ext-zopfli) one instead.

File: Dockerfile [hunk 3]
```diff
@@ -23,6 +23,7 @@ ENV PHP_SNAPPY_VERSION="bfefe4906e0abb1f6cc19005b35f9af5240d9025"
 ENV PHP_LZ4_VERSION="2f006c3e4f1fb3a60d2656fc164f9ba26b71e995"
 ENV PHP_XZ_VERSION=5.2.7
 ENV PHP_EXT_XZ_VERSION=1.1.2
+ENV PHP_ZOPFLI_VERSION="master"
```
Comment (stnguyen90) [inline Dockerfile:26]: Let's use a specific commit hash instead of master.  LIke this how we do brotli.

**Review by stnguyen90 — CHANGES_REQUESTED** [review]
Great PR! 🤯 We left some comments during the review, please check them out.

It also looks like the linter job is failing. Please make sure that passes by formatting the code.

Also, I tried to build the docker image but it failed:

image

File: Dockerfile (hunk 2 above)
Comment (toukirkhan) [reply Dockerfile]: I have made the changes with the suggested one. @stnguyen90

File: Dockerfile (hunk 3 above)
Comment (toukirkhan) [reply Dockerfile]: Used the commit hash for the latest commit in extension's repository @stnguyen90

**Comment by toukirkhan** [comment]
@stnguyen90 I have tried almost everything but I don't know why the "Zlib.h" header couldn't be found and hence, the build is failing. Can you please help me?

File: Dockerfile [hunk 4]
```diff
@@ -31,7 +32,8 @@ RUN apk add --no-cache \
   g++  \
   zstd-dev \
   brotli-dev \
-  lz4-dev
+  lz4-dev \
+  zopfli-dev
```
Comment (stnguyen90) [inline Dockerfile:36]: This should be:

```suggestion
  zlib-dev
```

File: Dockerfile [hunk 5]
```diff
@@ -86,6 +88,17 @@ RUN git clone https://github.com/codemasher/php-ext-xz.git --branch ${PHP_EXT_XZ
   && phpize \
   && ./configure \
   && make && make install
+  
+## Zopfli Extension
+FROM compile AS zopfli
+RUN git clone --recursive https://github.com/kjdev/php-ext-zopfli.git \
+  && cd php-ext-zopfli \
+  &&sudo apt-get update \
+  &&sudo apt-get install zlib1g-dev \
```
Comment (stnguyen90) [inline Dockerfile:97]: 1. This is not how packages are installed in alpine
2. Required packages should be installed in the `RUN apk add` section above

```suggestion
```

File: src/Storage/Compression/Algorithms/Zopfli.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
+@@ -0,0 +1,40 @@
```
Comment (stnguyen90) [inline src/Storage/Compression/Algorithms/Zopfli.php:1]: This shouldn't be here

```suggestion
```

File: src/Storage/Compression/Algorithms/Zopfli.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    /**
+     * Decompress.
+     *
+     * @param string $data
+     *
+     * @return string
+     */
+    public function decompress(string $data): string
+    {
+        return \zopflidecode($data);
```
Comment (stnguyen90) [inline src/Storage/Compression/Algorithms/Zopfli.php:39]: These are not the correct encode/decode functions for the new extension repo.

File: tests/Storage/Compression/Algorithms/ZopfliTest.php [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        $this->object = new Zopfli();
+    }
+
+    public function tearDown(): void
+    {
+    }
+
+    public function testName()
+    {
+        $this->assertEquals($this->object->getName(), 'zopfli');
```
Comment (stnguyen90) [inline tests/Storage/Compression/Algorithms/ZopfliTest.php:23]: The first parameter for `assertEquals()` should be the expected value and the 2nd should be the actual.

File: tests/Storage/Compression/Algorithms/ZopfliTest.php [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+        $this->assertEquals($demoSize, 3038056);
+        $this->assertEquals($dataSize, 2981000);
+
+        $this->assertGreaterThan($dataSize, $demoSize);
+
+        $data = $this->object->decompress($data);
+        $dataSize = mb_strlen($data, '8bit');
+
+        $this->assertEquals($dataSize, 3038056);
```
Comment (stnguyen90) [inline tests/Storage/Compression/Algorithms/ZopfliTest.php:75]: The values for the tests are incorrect so the tests fail.

**Review by stnguyen90 — CHANGES_REQUESTED** [review]
I've left suggestions.

Please also make sure to run the formatter so that the linter passes.

File: src/Storage/Compression/Algorithms/Zopfli.php (hunk 7 above)
Comment (toukirkhan) [reply src/Storage/Compression/Algorithms/Zopfli.php:38]: changing this one!

File: tests/Storage/Compression/Algorithms/ZopfliTest.php (hunk 8 above)
Comment (toukirkhan) [reply tests/Storage/Compression/Algorithms/ZopfliTest.php:23]: okay I got it!

File: tests/Storage/Compression/Algorithms/ZopfliTest.php (hunk 9 above)
Comment (toukirkhan) [reply tests/Storage/Compression/Algorithms/ZopfliTest.php:75]: I get it!

File: src/Storage/Compression/Algorithms/Zopfli.php (hunk 7 above)
Comment (stnguyen90) [reply src/Storage/Compression/Algorithms/Zopfli.php:38]: This hasn't been updated.

Comment (toukirkhan) [reply src/Storage/Compression/Algorithms/Zopfli.php]: Updated now

**Comment by stnguyen90** [comment]
@toukirkhan it looks like the linter and unit tests are still failing. Would you please address those?

When you're finished with your changes, please re-request a review from me so I know it's ready to be reviewed again.

**Pull request closed without merging**